src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java

```diff
@@ -0,0 +1,672 @@
+    /**
+     * Returns index % 6 to determine column
+     */
+    public int getColumn()
+    {
+        return (getTabIndex() > 11 ? (getTabIndex() - 12) % 12 : this.getTabIndex()) % 6;
+    }
+
+    /**
+     * @return Whether this tab's parent GUI is the current GUI
+     */
+    public boolean isActiveTab(Class<? extends GuiContainer> guiContainer)
+    {
+        return guiContainer == getTargetGui();
+    }
+
+    /**
+     * @return The tabindex / 12
+     */
+    public int getTabPage()
+    {
+        return getTabIndex() / 12;
+    }
+
+    /**
+     * @return The localized name for this tab
+     */
+    public String getLocalizedName()
+    {
+        return "guiGroup." + name;
+    }
+
+    public ResourceLocation getIconResLoc()
+    {
+        return iconResLoc;
+    }
+
+    public ItemStack getIconStack()
+    {
+        return iconStack;
+    }
+
+    /**
+     * @return A list of the appended tabs for the given GuiContainer
+     */
+    public static List<GuiTab> getTabsForGui(Class<? extends GuiContainer> guiContainer)
+    {
+        return tabRegistry.get(guiContainer);
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainer
+     *
+     * @param parentContainer The GuiContainer to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer> parentContainer)
+    {
+        if (parentContainer == GuiContainerCreative.class)
+        {
+            throw new IllegalArgumentException("Cannot add tabs to creative inventory. Use CreativeTabs instead");
+        }
+        else if (parentContainer == InventoryEffectRenderer.class)
+        {
+            throw new IllegalArgumentException(name + " tab has a parent that is ambiguous with the creative inventory. cannot proceed with registry.");
+        }
+
+        this.parentContainer = parentContainer;
+        tabRegistry.put(parentContainer, this);
+        return this;
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainers
+     *
+     * @param parentContainers The GuiContainers to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer>... parentContainers)
+    {
+        Arrays.stream(parentContainers).forEach(this::addTo);
+        return this;
+    }
+
+    /**
+     * Sets the GuiContainer this tab should show on as the main tab.
+     *
+     * @param targetGui The target gui to shown on
+     * @return The GuiTab instance
+     */
+    public GuiTab setTargetGui(Class<? extends GuiContainer> targetGui)
+    {
+        this.targetGui = targetGui;
+        return this;
+    }
+
+    /**
+     * Process interaction when the tab is clicked. should be used to send a packet to the server and open a GUI
+     *
+     * @param guiContainer The parent gui
+     */
+    public void onTabClicked(GuiContainer guiContainer)
+    {
+        int size = getTabsForGui(guiContainer.getClass()).size();
+        if (size > 0)
+        {
+            IntStream.range(0, size).filter(i ->
```

Comment:
change this, again/

Reply:
I can't be bothered to look up what this was originally, but this is really obscure.
Alternative swould be:
```java
List<GuiTab> targetTabs = getTabsForGui(targetGui);
List<GuiTab> tabs = getTabsForGui(guiContainer.getClass());
for (GuiTab tab : tabs)
{
    if (!targetTabs.contains(tab))
    {
        tab.addTo(targetGui);
    }
}
```

or if you really want to use streams, which imo is not a lot clearer due to the inversion (stupid Java not having `filterNot`):
```java
List<GuiTab> targetTabs = getTabsForGui(targetGui);
List<GuiTab> tabs = getTabsForGui(guiContainer.getClass());
tabs.stream().filter(Predicates.not(targetTabs::contains)).forEach(t -> t.addTo(targetGui));
```

Reply:
Using integers for index is absolutely the worst. Both solutions by @PaleoCrafter are way better.